File: 940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int left=0;
        int maxLength=0;
        unordered_map<int,int>mp;
        for(int right=0;right<fruits.size();right++){
            mp[fruits[right]]++;


            while(mp.size() > 2){
                mp[fruits[left]]--;
                if(mp[fruits[left]]==0){
                    mp.erase(fruits[left]);
                }
                left++;
            }

            maxLength=max(maxLength,right-left+1);
        }
        return maxLength;
    }
};

File: 940-fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int n=fruits.length;
        HashMap<Integer,Integer>mp=new HashMap<>();
        int left=0;
        int maxLength=0;
        for(int right=0;right<n;right++){
            mp.put(fruits[right],mp.getOrDefault(fruits[right],0)+1);
            while(mp.size() >2){
                mp.put(fruits[left], mp.get(fruits[left]) - 1);
                if(mp.get(fruits[left])==0){
                    mp.remove(fruits[left]);
                }
                left++;
            }

            maxLength=Math.max(maxLength,right-left+1);
        }

        return maxLength;
    }
}